app/src/main/cpp/skyline/common/circular_queue.h

```diff
@@ -77,46 +78,50 @@ namespace skyline {
         }
 
         Type Pop() {
-            std::unique_lock lock(productionMutex);
-            produceCondition.wait(lock, [this]() { return start != end; });
+            {
+                std::unique_lock productionLock{productionMutex};
+                produceCondition.wait(productionLock, [this]() { return start != end; });
+            }
 
+            std::scoped_lock comsumptionLock{consumptionMutex};
             auto next{start + 1};
             next = (next == reinterpret_cast<Type *>(vector.end().base())) ? reinterpret_cast<Type *>(vector.begin().base()) : next;
             Type item{*next};
             start = next;
 
-            if (start == end)
-                consumeCondition.notify_one();
+            consumeCondition.notify_one();
 
             return item;
         }
 
         void Push(const Type &item) {
-            std::unique_lock lock(productionMutex);
-            auto next{end + 1};
-            next = (next == reinterpret_cast<Type *>(vector.end().base())) ? reinterpret_cast<Type *>(vector.begin().base()) : next;
-            if (next == start) {
-                std::unique_lock consumeLock(consumptionMutex);
-                consumeCondition.wait(consumeLock, [=]() { return next != start; });
-            }
-            *next = item;
-            end = next;
-            produceCondition.notify_one();
-        }
+            Type *waitNext{};
 
-        void Append(span <Type> buffer) {
-            std::unique_lock lock(productionMutex);
-            for (const auto &item : buffer) {
+            while (true) {
+                if (waitNext) {
+                    std::unique_lock consumeLock{consumptionMutex};
+                    consumeCondition.wait(consumeLock, [=]() { return waitNext != start; });
+                    waitNext = nullptr;
+                }
+
+                std::scoped_lock lock{productionMutex};
                 auto next{end + 1};
                 next = (next == reinterpret_cast<Type *>(vector.end().base())) ? reinterpret_cast<Type *>(vector.begin().base()) : next;
                 if (next == start) {
-                    std::unique_lock consumeLock(consumptionMutex);
-                    consumeCondition.wait(consumeLock, [=]() { return next != start; });
+                    waitNext = next;
+                    continue;
                 }
                 *next = item;
-                end = next++;
+                end = next;
+                produceCondition.notify_one();
+                break;
             }
-            produceCondition.notify_one();
+
+        }
+
+        void Append(span<Type> buffer) {
+            for (const auto &item : buffer)
+                Push(item);
         }
```

Comment:
Previously `Append` was guaranteed to append elements in `buffer` in a contiguous manner, but that's not the case anymore. I wonder if that's a useful requirement, do we need it at all? E.g. one thread inserts `A,B,C` with `Append` and another one pushes `D` with `Push`, the final order would be `A,B,C,D` before, but now it's not guaranteed.

In any case, a comment stating this explicitly would make sense.

Reply:
That's fine, I don't think that was ever a particular guarantee but ye will mention